silence/silence.go

```diff
@@ -611,43 +602,55 @@ func (s *Silences) Set(sil *pb.Silence) error {
 	defer s.mtx.Unlock()
 
 	now := s.nowUTC()
-	prev, ok := s.getSilence(sil.Id)
+	if sil.StartsAt.IsZero() {
+		sil.StartsAt = now
+	}
+
+	if err := validateSilence(sil); err != nil {
+		return fmt.Errorf("invalid silence: %w", err)
+	}
+
+	existing, ok := s.getSilence(sil.Id)
 	if sil.Id != "" && !ok {
 		return ErrNotFound
 	}
 
-	if ok && canUpdate(prev, sil, now) {
-		return s.setSilence(sil, now, false)
+	// updateExisting is true iff this is an existing silence, and the update does
```

Comment:
In the end I kept all this code. Here is the reasoning:

1. I've done a lot of reasoning and manual testing, in additional to the unit tests, and I'm more confident about it then other implements I tried in the last week. This implementation is the outcome of about 5 hours of thinking and experimenting how I can solve this problem as cleanly as possible.
2. I think now that we've made the fixes more incremental it's easier to review then it was in #3892.

Please let me know if this is a problem and I'll see how I can rewrite the code to be more like the original.

Reply:
Here is an alternative implementation which you might prefer:

```diff
@@ -611,43 +602,58 @@ func (s *Silences) Set(sil *pb.Silence) error {
        defer s.mtx.Unlock()

        now := s.nowUTC()
-       prev, ok := s.getSilence(sil.Id)
-       if sil.Id != "" && !ok {
-               return ErrNotFound
+       if sil.StartsAt.IsZero() {
+               sil.StartsAt = now
        }

-       if ok && canUpdate(prev, sil, now) {
-               return s.setSilence(sil, now, false)
+       if err := validateSilence(sil); err != nil {
+               return fmt.Errorf("invalid silence: %w", err)
        }

-       // If we got here it's either a new silence or a replacing one (which would
-       // also create a new silence) so we need to make sure we have capacity for
-       // the new silence.
-       if s.limits.MaxSilences != nil {
-               if m := s.limits.MaxSilences(); m > 0 && len(s.st)+1 > m {
-                       return fmt.Errorf("exceeded maximum number of silences: %d (limit: %d)", len(s.st), m)
-               }
+       existing, ok := s.getSilence(sil.Id)
+       if sil.Id != "" && !ok {
+               return ErrNotFound
        }

-       if ok && getState(prev, s.nowUTC()) != types.SilenceStateExpired {
-               // We cannot update the silence, expire the old one to leave a history of
-               // the silence before modification.
-               if err := s.expire(prev.Id); err != nil {
-                       return fmt.Errorf("expire previous silence: %w", err)
+       if ok && canUpdate(existing, sil, now) {
+               sil.UpdatedAt = now
+               msil := s.toMeshSilence(sil)
+               if err := s.checkLimits(msil, true); err != nil {
+                       return err
                }
+               return s.setSilence(msil, now)
        }

+       // If this is a new silence, or it is replacing an existing silence,
+       // then give it an ID and make sure the start time is not in the past.
        uid, err := uuid.NewV4()
        if err != nil {
-               return fmt.Errorf("generate uuid: %w", err)
+               return fmt.Errorf("failed to create UUID: %w", err)
        }
        sil.Id = uid.String()
-
        if sil.StartsAt.Before(now) {
                sil.StartsAt = now
        }
+       sil.UpdatedAt = now

-       return s.setSilence(sil, now, false)
+       // Check the limits before expiring the existing silence to prevent
+       // partial updates when limits are exceeded. Limits are checked against
+       // the mesh silence as this is the silence that is gossiped and stored
+       // on disk.
+       msil := s.toMeshSilence(sil)
+       if err := s.checkLimits(msil, false); err != nil {
+               return err
+       }
+
+       if ok && getState(existing, s.nowUTC()) != types.SilenceStateExpired {
+               // We cannot update the silence, expire the old one to leave a history of
+               // the silence before modification.
+               if err := s.expire(existing.Id); err != nil {
+                       return fmt.Errorf("expire previous silence: %w", err)
+               }
+       }
+
+       return s.setSilence(msil, now)
 }
```